File: data_structure/HeavyLightDecomposition/HLD_base.cpp
struct HLD_base {
	int n, tot, fa[MAXN], depth[MAXN], siz[MAXN], son[MAXN], id[MAXN], top[MAXN], ori[MAXN];

	void Init(const GRAPH& g, int rt) {
		n = g.s.size()-1;
		tot = 0;
		memset(son, 0, (n+1) * sizeof(int));
		dfs1(g, rt, 0, 1);
		dfs2(g, rt, rt);
	}
	void dfs1(const GRAPH& g, int s, int f, int d) {
		depth[s] = d;
		fa[s] = f;
		siz[s] = 1;
		for (int v : g.s[s]) {
			if (v == f) continue;
			dfs1(g, v, s, d+1);
			siz[s] += siz[v];
			if (siz[v] > siz[son[s]])
				son[s] = v;
		}
	}
	//t is the top of the current list
	void dfs2(const GRAPH& g, int s, int t) { 
		top[s] = t;
		id[s] = ++tot;
		ori[tot] = s;
		if (!son[s]) return;
		dfs2(g, son[s], t);
		for (int v : g.s[s])
			if (v != son[s] && v != fa[s])
				dfs2(g, v, v);
	}
};

File: rust/data_structure/HLD/HLDBase.rs
struct HLDBase {
    n: usize,
    fa: Vec<usize>,
    depth: Vec<usize>,
    siz: Vec<usize>,
    id: Vec<usize>,
    top: Vec<usize>,
    ori: Vec<usize>,
}
impl HLDBase {
    fn dfs1(&mut self, g: &Graph, son: &mut Vec<usize>, s: usize, f: usize, d: usize) {
        self.depth[s] = d;
        self.fa[s] = f;
        self.siz[s] = 1;
        for v in &g.s[s] {
            let v = *v;
            if v == f {
                continue;
            }
            self.dfs1(g, son, v, s, d + 1);
            self.siz[s] += self.siz[v];
            if self.siz[v] > self.siz[son[s]] {
                son[s] = v;
            }
        }
    }
    //t is the top of the current list
    fn dfs2(&mut self, g: &Graph, son: &Vec<usize>, tot: &mut usize, s: usize, t: usize) {
        self.top[s] = t;
        *tot += 1;
        self.id[s] = *tot;
        self.ori[*tot] = s;
        if son[s] == 0 {
            return;
        }
        self.dfs2(g, son, tot, son[s], t);
        for v in &g.s[s] {
            let v = *v;
            if v != son[s] && v != self.fa[s] {
                self.dfs2(g, son, tot, v, v);
            }
        }
    }
    fn new(g: &Graph, rt: usize) -> Self {
        let n = g.s.len() - 1;
        let mut tot = 0;
        let mut son = vec![0; n+1];
        // Actually no need to be 0
        let mut hld = Self {
            n,
            fa: vec![0; n+1],
            depth: vec![0; n+1],
            siz: vec![0; n+1],
            id: vec![0; n+1],
            top: vec![0; n+1],
            ori: vec![0; n+1],
        };
        hld.dfs1(g, &mut son, rt, 0, 1);
        hld.dfs2(g, &son, &mut tot, rt, rt);
        hld
    }
}
